sql: add open txns gauge

jordanlewis opened the pull request:
This is a new gauge in `sql.txns.open` that tracks the number of open
transactions.

Closes #57475
Closes #52005

cc @a-entin @awoods187 

The dashboard now contains the following 3 graphs first:

1. Open SQL Sessions
2. Open SQL Transactions
3. Active SQL Queries

Release note (ui change): add Open SQL Transactions to SQL dashboard in
DB Console. Rename SQL Queries to SQL Statements in SQL metrics
dashboard. Remove references to "Distributed SQL Queries" when metrics
really are discussing all SQL queries.

jordanlewis committed:
sql: add open txns gauge

This is a new gauge in `sql.txns.open` that tracks the number of open
transactions.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57477)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
This is not exposed in the UI right now, right? Do we want a dashboard for it? If not, how is it accessible for users/SREs?

The code <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57477#-:-MNdmTl87IEfL1Wd1v2q:bd1jfhg)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

awoods187 commented:
We definitely should add this to the DB console too.

jordanlewis committed:
ui: add Open SQL Transactions; reorder dashboard

The dashboard now contains the following 3 graphs first:

1. Open SQL Sessions
2. Open SQL Transactions
3. Active SQL Queries

Release note (ui change): add Open SQL Transactions to SQL dashboard in
DB Console. Rename SQL Queries to SQL Statements in SQL metrics
dashboard. Remove references to "Distributed SQL Queries" when metrics
really are discussing all SQL queries.

jordanlewis commented:
Added to the DB Console. I also took the liberty of homogenizing the DB Console's metrics in this department for clarity.

The top 3 metrics are now:

1. Open SQL Connections (renamed from SQL Connections)
2. Open SQL Transactions (the new metric)
3. Active SQL Queries (renamed from Active Distributed SQL Queries)

![image](https://user-images.githubusercontent.com/43821/101108182-fa489680-35a1-11eb-8756-d79ba2704e7a.png)

yuzefovich reviewed (approved):
> Reviewed 4 of 4 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57477)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

rafiss commented on pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx:
```diff
@@ -40,6 +40,26 @@ export default function (props: GraphDashboardProps) {
+      tooltip={`The total number of SQL queries currently running ${tooltipSelection}.`}
+    >
+      <Axis label="queries">
+        <Metric name="cr.node.sql.distsql.queries.active" title="Active Queries" />
```
drive-by comment -- does this metric capture all queries (not just distSQL)? is that new or the same?

should the comment on the `sql.distsql.queries.active` metric be updated in execinfra/metrics.go to say this?

awoods187 replied:
Yeah im curious on both of these points too

jordanlewis replied:
All queries, yes

rafiss commented on pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx:
```diff
@@ -40,6 +40,26 @@ export default function (props: GraphDashboardProps) {
+      tooltip={`The total number of SQL transactions currently open ${tooltipSelection}.`}
+    >
+      <Axis label="transactions">
+        <Metric name="cr.node.sql.txns.open" title="Open Transactions" />
```
drive-by comment -- i'm curious if this includes open implicit transactions?

more fundamentally, i'm interested in how would i answer the question: "how many of my SQL connections are currently doing something?" which metric would i use for that? (no prob if the answer is none of these, i'm just trying to understand these metrics, not suggesting any changes)

jordanlewis replied:
It does include implicit txns

jordanlewis replied:
I think "doing something" would be satisfied by open txns, which cost resources even if the connection is idle. Maybe active queries would be slightly better, since those are doing "more work".

rafiss commented:
is there already a process for getting the new DB console chart documented here: https://www.cockroachlabs.com/docs/v20.2/ui-sql-dashboard.html (curious if this should be in the Release Note?)

a-entin commented:
@jordanlewis and @awoods187 : one though is we use both terms - "connections" and "sessions" in the UI. While they are synonymous, it maybe better to pick one for consistency. I don't have a strong opinion about it since say ODBC/JDBC vernacular is connections and in our database users manage sessions. Yet it maybe worthwhile to define and perhaps even put in the doc a nice taxonomy, then revisit all user visible labels everywhere.

awoods187 commented:
> While they are synonymous, it may be better to pick one for consistency. 
I agree. I like the idea of naming this metric `open sessions` to match the sessions page. We can revisit the decision on sessions vs. connections in separate work.

For similar reasons, I think the last chart should be renamed from `queries` to `statements`.

I also like the hierarchy of sessions->transactions->statements as it matches the left-hand nav on those individual pages..

a-entin commented:
I second Andy's suggestion.

A nit... isn't "open" implied in "open sessions"? What other sessions could there be. Can be omitted. Yet a clarification re user (or "client") sessions / transactions may be helpful to reflect if can be 1 adjective added to a label.  This is merely a thought, should in no way delay getting the essential new metric into the UI.

awoods187 commented:
I think we should backport this to 20.2.X as well.

jordanlewis commented:
Open isn't implied. A metric could count the cumulative number of sessions that have been opened ever.

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57477#-:-MQnn2TTA3XzkOoPD5QY:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss and @yuzefovich)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
bors r+
